### PR TITLE
[IMP] mail: EmojiPicker in mobile mode

### PR DESCRIPTION
**Pull request by zel-odoo**
task-3015707

It updated the way to display `EmojiPicker` in mobile mode.

Before: 
image

After:

---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/mail/static/src/composer/composer.js [hunk 1]
```diff
@@ -128,10 +130,23 @@ export class Composer extends Component {
...
+                scrollValue: 0,
+                set: (value) => {
+                    this.storeScrollEmojiPicker.scrollValue = value;
+                },
+                get: () => {
+                    return this.storeScrollEmojiPicker.scrollValue;
+                },
+            };
+        }
+        onExternalClick("composer", () => (this.state.showEmojiPicker = false));
```
[the remark this answers is not in the document]
Comment (alexkuhn) [reply addons/mail/static/src/composer/composer.js:149]: For emoji picker this makes sense to have it.
- Most pPopovers are expected to close on click away
- In mobile, this is an extension of soft keyboard, and users expect to remove soft keyboard when removing the focus on the input (generally with click away)

**Comment by alexkuhn** [comment]
Emoji picker is too small
Screenshot 2023-05-10 at 13 59 35

I would expect composer + emoij picker to take 50% of screen.
Also category bar could be moved at the bottom, so it's less important than the main content of emoij picker.
There'll be a small issue of having 2 input next to each other. We should improve design of pickers, as explained in this comment:
https://github.com/odoo/odoo/pull/116060#issuecomment-1542074463

In short:
- Having a bar to select which picker (Gif or emoji)
- Show emoji category bar vertically in popover, in order to not bloat header in popover configuration

File: addons/mail/static/src/composer/composer.xml [hunk 2]
```diff
@@ -100,6 +100,7 @@
                 </div>
             </div>
         </div>
+        <EmojiPicker t-if="env.isSmall and env.inChatWindow" onSelect="(str) => this.addEmoji(str)" close="() => this.state.autofocus++" storeScroll="storeScrollEmojiPicker" className="state.showEmojiPicker ? 'o-show' : 'o-hide'"/>
```
Comment (phenix-factory) [inline addons/mail/static/src/composer/composer.xml:103]: The `env.inChatWindow` doesn't seams useful since you already test `isSmall`.

File: addons/mail/static/src/emoji_picker/emoji_picker.xml [hunk 3]
```diff
@@ -2,7 +2,7 @@
 <templates xml:space="preserve">
 
 <t t-name="mail.EmojiPicker" owl="1">
-    <div class="o-mail-EmojiPicker bg-white" t-att-class="emojis.length === 0 ? 'd-flex flex-column align-items-center justify-content-center' : ''" t-on-click="onClick" t-on-keydown="onKeydown">
+    <div class="o-mail-EmojiPicker bg-white" t-attf-class=" {{props.className}} {{emojis.length === 0 ? 'd-flex flex-column align-items-center justify-content-center' : ''}} {{env.isSmall and env.inChatWindow ? 'o-in-small' : ''}}" t-on-click="onClick" t-on-keydown="onKeydown">
```
Comment (alexkuhn) [inline addons/mail/static/src/emoji_picker/emoji_picker.xml:5]: `o-in-small` => `o-small` (avoid modifiers with dash separator on many words. use camelCase

File: addons/mail/static/src/emoji_picker/emoji_picker.xml [hunk 4]
```diff
@@ -18,7 +18,7 @@
                     </t>
                 </t>
             </div>
-            <div class="p-1 d-flex flex-column">
+            <div class="o-mail-EmojiPicker-searchBar p-1 d-flex flex-column">
```
Comment (alexkuhn) [inline addons/mail/static/src/emoji_picker/emoji_picker.xml:21]: `searchBar` => `search` (no need for bar, and it's easier to read when camelCase is avoidable)

File: addons/mail/static/src/composer/composer.js (hunk 1 above)
Comment (alexkuhn) [inline addons/mail/static/src/composer/composer.js:149]: As we discussed, it's no good to have 2 ways to use EmojiPicker (either hook or as sub-component in template).
This leads to lots of code duplication, and also it makes `composer` more complex than it should.

I would keep a single way to `useEmojiPicker` with the hook, and let the hook allow to use emoji picker with popover or non-popover based on params you pass to the hook.

**Review by alexkuhn — CHANGES_REQUESTED** [review]

File: addons/mail/static/src/emoji_picker/emoji_picker.scss [hunk 5]
```diff
@@ -3,8 +3,20 @@
...
+    "search-bar"
+    "content";
     grid-template-rows: 34px 35px 1fr;
 
+    &.o-in-small {
+        grid-template-areas:
+        "search-bar"
+        "content"
+        "header" !important;
+        grid-template-rows: 35px 1fr 34px !important;
```
Comment (alexkuhn) [inline addons/mail/static/src/emoji_picker/emoji_picker.scss:17]: part name can just be `search` here too

File: addons/mail/static/src/composer/composer.xml [hunk 6]
```diff
@@ -90,6 +90,7 @@
                     imagesHeight="50"/>
             </div>
         </div>
+        <EmojiPicker t-if="env.isSmall" onSelect="(str) => this.addEmoji(str)" close="() => this.state.autofocus++" restorePositionOnSelect="false" storeScroll="useEmojiPickerStoreScroll()" className="state.showEmojiPicker ? 'o-show' : 'o-hide'"/>
```
Comment (alexkuhn) [inline addons/mail/static/src/composer/composer.xml:93]: We should always mount the `EmojiPicker` and have it handle its visibility + internal state `storeScroll`.
By doing so, you won't have to define the `useEmojiPickerStoreScroll`, and you won't have to rely on `d-none` in the concern of Composer (this is an implementation detail of the `EmojiPicker`, and actually you could do this without `d-none`)

File: addons/mail/static/src/emoji_picker/emoji_picker.js [hunk 7]
```diff
@@ -20,6 +20,19 @@ import { usePopover } from "@web/core/popover/popover_hook";
...
+        scrollValue: 0,
+        set: (value) => {
+            storeScroll.scrollValue = value;
+        },
+        get: () => {
+            return storeScroll.scrollValue;
+        },
+    };
+    return storeScroll;
+}
```
Comment (alexkuhn) [inline addons/mail/static/src/emoji_picker/emoji_picker.js:34]: This function has no reason to exists. It should be internal code of the `Emoji Picker`, as I said in previous comment

File: addons/mail/static/src/emoji_picker/emoji_picker.js [hunk 8]
```diff
@@ -121,8 +126,15 @@ export async function loadEmoji() {
...
-    static props = ["onSelect", "close", "onClose?", "storeScroll?"];
-    static defaultProps = { onClose: () => {} };
+    static props = [
+        "onSelect",
+        "close",
+        "onClose?",
+        "restorePositionOnSelect?",
+        "storeScroll?",
+        "className?",
+    ];
```
Comment (alexkuhn) [inline addons/mail/static/src/emoji_picker/emoji_picker.js:136]: I'd expect the list of props to not change in PR. At most just a single prop to manage visibility, but even that I'm not entirely sure it's needed. `storeScroll` should be internal code of `EmojiPicker`.

File: addons/mail/static/src/emoji_picker/emoji_picker.js [hunk 9]
```diff
@@ -294,7 +306,9 @@ export class EmojiPicker extends Component {
         this.recent[codepoints] ??= 0;
         this.recent[codepoints]++;
         browser.localStorage.setItem("mail.emoji.frequent", JSON.stringify(this.recent));
-        this.gridRef.el.scrollTop = 0;
+        if (this.props.restorePositionOnSelect && !ev.shiftKey) {
+            this.gridRef.el.scrollTop = 0;
+        }
```
Comment (alexkuhn) [inline addons/mail/static/src/emoji_picker/emoji_picker.js:311]: I don't see the point of `restorePositionOnSelect` prop: why not reset scrollposition in mobile? It feels natural to reset in that context too (Discord does that btw)

**Review by alexkuhn — CHANGES_REQUESTED** [review]
The header and content of emoji picker feels too small for mobile. I suggest bigger emojis:
Screenshot 2023-05-31 at 13 20 41

File: addons/mail/static/src/composer/composer.xml (hunk 6 above)
Comment (zel-odoo) [reply addons/mail/static/src/composer/composer.xml:93]: Yes, but the problem is when we are using popover for the `EmojiPicker`, the component is always created and destroyed. This is the behavior of the popover hook. If we always have a popover but replace everything related to `onClose` with `onClick` and make the popover hidden when `onClick`, it works, but not the idea of how a component should be using a popover hook. that's why we need to store the position outside of the component.

File: addons/mail/static/src/emoji_picker/emoji_picker.js (hunk 9 above)
Comment (zel-odoo) [reply addons/mail/static/src/emoji_picker/emoji_picker.js:311]: No, in mobile, when you click once, the EmojiPicker will not reset the scroll position. It only resets when it closes. This behaves differently on mobile and desktop. On the desktop, once we click it, the popover is gone, we can only send one by one. But in mobile, we can do it multiple times, and we don't want to restore the positon after inserting each emoji.

Comment (alexkuhn) [reply addons/mail/static/src/emoji_picker/emoji_picker.js:311]: Oh I see, when using emoji picker in composer, but not for reactions.
This code should be internally of EmojiPicker. This resets should happen the emoji picker is closed, it should have nothing to do with selection (which is one way to close the emoji picker, but it shouldn't be explicitly coupled with selection. Instead it's the closing behaviour of emoji picker)

File: addons/mail/static/src/composer/composer.xml (hunk 6 above)
Comment (alexkuhn) [reply addons/mail/static/src/composer/composer.xml:93]: I really don't like how some code that is internal of emoji picker because bits outside of internal code of emoji picker.

Sure, we tricked by putting state in the component hook rather than `EmojiPicker`, and this helps having persistence of state when emoji picker is closed. That's the nice thing of component hook: it defines a feature rather than a component, so the feature can work without coupling it to sub-child lifecycle.

Component hooks have their limitations, so I get why you use component... But I think components that are pure showing of content like how `EmojiPicker` currently is, this prevents encapsulate all the logic of the feature, including when the content is not shown.

I really think this case should also work with `useEmojiPicker()`. I get it's not conventional to do this, but it's one of the 1st component where we want 2 different presentation in desktop/mobile, and I really dislike this code duplication + totally different approaches when it's just a change of container + leaking of internal features of EmojiPicker + added complexity of API of `EmojiPicker`.

The increase of complexity should stay inside the EmojiPicker, not outside.
Using the EmojiPicker should stays very straightforward to use.

**Comment by zel-odoo** [comment]
@alexkuhn According to your advice, I will try to refactor the `EmojiPicker` code and create `Picker` or `MobileDesktopSwitcher` or something similar. I will keep this PR open and we will see if that PR is making the code better or not. I will send the link when having a code.

**Comment by alexkuhn** [comment]
Ok, let's see whether this will improve code of EmojiPicker when considering mobile view. Thanks!

**Comment by zel-odoo** [comment]
https://github.com/odoo/odoo/pull/123671 I create a branch and the changes we mentioned about the component hook will be there.
